hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java

```diff
@@ -0,0 +1,174 @@
+import java.util.Map;
+
+import org.apache.hudi.client.utils.LazyIterableIterator;
+import org.apache.hudi.common.data.HoodieData;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.index.HoodieIndex;
+import org.apache.hudi.table.WorkloadProfile;
+import org.apache.hudi.table.action.commit.Partitioner;
+import org.apache.hudi.table.action.commit.SparkBucketIndexPartitioner;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import org.apache.hudi.client.WriteStatus;
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordLocation;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieIndexException;
+import org.apache.hudi.index.HoodieIndexUtils;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.spark.api.java.JavaRDD;
+
+public class SparkBucketIndex<T extends HoodieRecordPayload>
+    extends HoodieIndex<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> {
+
+  private static final Logger LOG =  LogManager.getLogger(SparkBucketIndex.class);
+
+  private final int numBuckets;
+
+  public SparkBucketIndex(HoodieWriteConfig config) {
+    super(config);
+    numBuckets = config.getBucketIndexNumBuckets();
+    LOG.info("use bucket index, numBuckets=" + numBuckets);
+  }
+
+  @Override
+  public HoodieData<WriteStatus> updateLocation(HoodieData<WriteStatus> writeStatuses,
+      HoodieEngineContext context,
+      HoodieTable hoodieTable)
+      throws HoodieIndexException {
+    return writeStatuses;
+  }
+
+  @Override
+  public HoodieData<HoodieRecord<T>> tagLocation(HoodieData<HoodieRecord<T>> records,
+      HoodieEngineContext context,
+      HoodieTable hoodieTable)
+      throws HoodieIndexException {
+    HoodieData<HoodieRecord<T>> taggedRecords = records.mapPartitions(recordIter -> {
+      // partitionPath -> bucketId -> fileInfo
+      Map<String, Map<Integer, Pair<String, String>>> partitionPathFileIDList = new HashMap<>();
+      return new LazyIterableIterator<HoodieRecord<T>, HoodieRecord<T>>(recordIter) {
+
+        @Override
+        protected void start() {
+
+        }
+
+        @Override
+        protected HoodieRecord<T> computeNext() {
+          HoodieRecord record = recordIter.next();
+          int bucketId = BucketIdentifier.getBucketId(record, config.getBucketIndexHashField(),
+              numBuckets, config.getBucketIndexHashFunction());
+          String partitionPath = record.getPartitionPath();
+          if (!partitionPathFileIDList.containsKey(partitionPath)) {
+            partitionPathFileIDList.put(partitionPath, loadPartitionBucketIdFileIdMapping(hoodieTable, partitionPath));
```

Comment:
rename: partitionPath

Reply:
don't get the point, can u explain this?